src/cli/upload.rs

```diff
@@ -33,9 +41,30 @@ impl UploadCommand {
     pub fn run(self) -> Result<(), anyhow::Error> {
         let project_path = resolve_path(&self.project);
 
-        let cookie = self.cookie.or_else(get_auth_cookie).context(
-            "Rojo could not find your Roblox auth cookie. Please pass one via --cookie.",
-        )?;
+        let use_open_cloud = self.api_key.is_some();
+
+        // Validate differently depending on if we're trying to use open cloud or not.
+        // If there's a better way of doing this, please do so.
+        let api_key = if use_open_cloud {
+            self.api_key
+                .context("Rojo could not find your api key. Please pass one via --api_key")?
+        } else {
+            "undefined".to_string()
+        };
+        let universe_id = if use_open_cloud {
+            self.universe_id.context(
+				"A Universe id is required when using the Open Cloud API. Please pass one via --universe_id"
+			)?
+        } else {
+            0
+        };
+        let cookie = if use_open_cloud {
+            "undefined".to_string()
+        } else {
+            self.cookie.or_else(get_auth_cookie).context(
+                "Rojo could not find your Roblox auth cookie. Please pass one via --cookie.",
+            )?
+        };
```

Comment:
This logic is pretty funky. I think we should move this logic down to where `do_upload` is called, and take advantage of Rust's match statement against all three of these values. You can use this to make some very detailed diagnostic messages for users, which can be very helpful!

I would write something like this:

```rust
match (self.cookie, self.api_key, self.universe_id) {
	(cookie, None, universe) => {
		if universe.is_some() {
			log::warn!("--universe_id was provided but is ignored when using legacy upload API");
		}

		let cookie = cookie.unwrap_or(get_auth_cookie).context("<good no cookie error>")?;
		do_upload(buffer, self.asset_id, &cookie)
	}

	(cookie, Some(api_key), Some(universe_id)) => {
		if cookie.is_some() {
			log::warn!("--cookie was provided but is ignored when using Open Cloud API");
		}

		do_upload_open_cloud(buffer, universe_id, self.asset_id, &api_key)
	}

	(_, Some(_), None) => {
		bail!("If --api_key is provided to upload, --universe_id must also be provided");
	}
}
```

Reply:
That's very close to what I was originally thinking (although I didn't know you could have multiple inputs in a match statement), but this would mean that it would have to build the buffer before it is able to verify if the information given is even correct. 

While this isn't technically an issue, I would like to avoid having that extra delay before you get to know if your information is correct. We may be able to keep the minimal delay by writing the binary file after it has verified the information but building the session and tree beforehand to minimize repeated code. (assuming writing the binary is actually what is taking the most amount of time)

I'll look into the best way to integrate this without adding unnecessary delay.